src/Domain/Neighbors.cpp

```diff
@@ -6,22 +6,31 @@
 #include <ostream>
 #include <pup.h>  // IWYU pragma: keep
 
-#include "Domain/ElementId.hpp"      // IWYU pragma: keep
+#include "Domain/ElementId.hpp"  // IWYU pragma: keep
+#include "Domain/MaxNumberOfNeighbors.hpp"
+#include "ErrorHandling/Assert.hpp"
 #include "Parallel/PupStlCpp11.hpp"  // IWYU pragma: keep
 #include "Utilities/GenerateInstantiations.hpp"
 #include "Utilities/StdHelpers.hpp"  // IWYU pragma: keep
 
 template <size_t VolumeDim>
 Neighbors<VolumeDim>::Neighbors(std::unordered_set<ElementId<VolumeDim>> ids,
                                 OrientationMap<VolumeDim> orientation) noexcept
-    : ids_(std::move(ids)), orientation_(std::move(orientation)) {}
+    : ids_(std::move(ids)), orientation_(std::move(orientation)) {
+  ASSERT(ids_.size() <= maximum_number_of_neighbors(VolumeDim),
```

Comment:
Actually, I think this is incorrect: `Neighbor` is the neighbors in just one particular direction, so the max is a smaller number (maximum_number_of_neighbors / number_of_faces)

Reply:
Yes, that is correct, this is for a particular direction.  Thus the smaller number is `2^(VolumeDim - 1)` assuming a maximum 2-to-1 refinement between two adjacent Elements (which `maximum_number_of_neighbors` assumes)

Reply:
I propose renaming `maximum_number_of_neighbors` to `maximum_number_of_element_neighbors`, and adding a new similar `maximum_number_of_face_neighbors`. Thoughts?

Reply:
I don't know what element and face neighbors means exactly, so I would prefer using Voronoi and Neumann neighbors as the terms since these actually exist in the literature. I think in either case the documentation needs to be clear. Voronoi is all elements that share a vertex with the self element. Neumann is all elements that share part or all of a face with the self element. If we need more functions we can be explicit, but at least these names would be consistent with what the literature calls them.

Reply:
It sounds like you are distinguishing between two different definitions for the neighbors of an element, whereas I am trying to distinguish between the neighbors of an element vs. the neighbors of one face of the element.

Another suggestion, then: `maximum_number_of_neighbors` (unchanged from current) vs. `maximum_number_of_neighbors_per_direction`